### PR TITLE
ofctrl library patch work

### PR DESCRIPTION
**Pull request by echkenluo**
1. Add pkt mark match and set support
2. Add load action support
3. Add resubmit action support
4. Add learn action support
5. Add empty forwarding elem

### CODE REVIEW

File: ofctrl/fgraphFlow.go [hunk 1]
```diff
@@ -212,8 +475,13 @@ func (self *Flow) xlateMatch() openflow13.Match {
...
 	if self.Match.VlanId != 0 {
-		vidField := openflow13.NewVlanIdField(self.Match.VlanId, nil)
-		ofMatch.AddField(*vidField)
+		if self.Match.VlanIdMask != nil {
+			vlanIdField := openflow13.NewVlanIdField(self.Match.VlanId, self.Match.VlanIdMask)
+			ofMatch.AddField(*vlanIdField)
+		} else {
+			vidField := openflow13.NewVlanIdField(self.Match.VlanId, nil)
+			ofMatch.AddField(*vidField)
+		}
```
Comment (KeyongSun) [inline ofctrl/fgraphFlow.go:484]: NewVlanIdField handled the vlanIDMask nil or not nil case, so no need to check VlanIdMask here.

File: ofctrl/fgraphFlow.go [hunk 2]
```diff
@@ -490,7 +779,55 @@ func (self *Flow) installFlowActions(flowMod *openflow13.FlowMod,
 			addActn = true
 
 			log.Debugf("flow install. Added setUDPDst Action: %+v", setUDPDstAction)
+		case "loadAction":
```
Comment (KeyongSun) [inline ofctrl/fgraphFlow.go:782]: Add a space between each case.

File: ofctrl/fgraphFlow.go [hunk 3]
```diff
@@ -490,7 +779,55 @@ func (self *Flow) installFlowActions(flowMod *openflow13.FlowMod,
...
+			loadAct := flowAction.loadAction
+			ofsBits := loadAct.Range.ToOfsBits()
+			loadOfAction := openflow13.NewNXActionRegLoad(ofsBits, loadAct.Field, loadAct.Value)
+
+			err := actInstr.AddAction(loadOfAction, true)
+			if err != nil {
+				return err
+			}
```
Comment (KeyongSun) [inline ofctrl/fgraphFlow.go:790]: It's better follow the style of this function implementation. "AddAction" seems will always return nil.

File: ofctrl/fgraphFlow.go [hunk 4]
```diff
@@ -556,6 +893,19 @@ func (self *Flow) install() error {
...
+	case "empty":
+		instr := self.NextElem.GetFlowInstr()
+		if instr != nil {
+			err := self.installFlowActions(flowMod, instr)
+			if err != nil {
+				return err
+			}
+			if len(instr.(*openflow13.InstrActions).Actions) > 0 {
+				flowMod.AddInstruction(instr)
+			}
```
Comment (KeyongSun) [inline ofctrl/fgraphFlow.go:905]: installFlowActions always return nil, it's better to follow the function impl style or add some debug log when err.

File: ofctrl/fgraphFlow.go (hunk 3 above)
Comment (echkenluo) [reply ofctrl/fgraphFlow.go:790]: done

File: ofctrl/fgraphFlow.go (hunk 1 above)
Comment (echkenluo) [reply ofctrl/fgraphFlow.go:484]: yes, I have check the VlanIdMask nil case. what's mean's that?

Comment (KeyongSun) [reply ofctrl/fgraphFlow.go:484]: 不需要if/else再判断 VlanIDMask 是不是nil：
可以直接调用：
vlanIdField := openflow13.NewVlanIdField(self.Match.VlanId, self.Match.VlanIdMask)
ofMatch.AddField(*vlanIdField)

**Review by KeyongSun — APPROVED** [review]

**Pull request merged**